### PR TITLE
sql: add missing features to support psqlODBC

### PR DESCRIPTION
**Pull request by benesch**
See individual commits for details. The upshot is that, with these additions, connecting Microsoft Excel to Materialize works at a basic level:

image

### Motivation

  * This PR adds a known-desirable feature: fixes #8645.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: src/pgrepr/src/types.rs [hunk 1]
```diff
@@ -91,6 +91,9 @@ pub enum Type {
         /// The type of the values in the map.
         value_type: Box<Type>,
     },
+    /// A character type for storing identifiers of no more than 64 characters
```
Comment (nrainer-materialize) [inline src/pgrepr/src/types.rs:94]: characters or bytes?

File: src/repr/src/strconv.rs [hunk 2]
```diff
@@ -606,6 +607,21 @@ where
...
+    for c in s.chars() {
+        len += c.len_utf8();
+        if len > NAME_MAX_BYTES - 1 {
+            break;
+        }
+        out.push(c);
+    }
+    out
+}
+
```
Comment (nrainer-materialize) [inline src/repr/src/strconv.rs:624]: Can we add a test for this function?

File: src/repr/src/strconv.rs [hunk 3]
```diff
@@ -606,6 +607,21 @@ where
     Nestable::MayNeedEscaping
 }
 
+pub fn parse_pg_legacy_name(s: &str) -> String {
```
Comment (nrainer-materialize) [inline src/repr/src/strconv.rs:610]: What do we do if we end up with collisions? Is this something we ignore for now?

File: src/pgrepr/src/types.rs (hunk 1 above)
Comment (jkosh44) [reply src/pgrepr/src/types.rs:94]: PostgreSQL defines this as a "Special Character Type" in their docs (https://www.postgresql.org/docs/15/datatype-character.html). So I think character type is probably accurate.

File: src/repr/src/strconv.rs (hunk 3 above)
Comment (jkosh44) [reply src/repr/src/strconv.rs:610]: I think this responsibility should be up to the caller of this function. We don't actually use this to parse object identifiers so I think it's fine.

File: src/repr/src/strconv.rs (hunk 2 above)
Comment (jkosh44) [reply src/repr/src/strconv.rs]: Done

File: src/pgrepr/src/types.rs (hunk 1 above)
Comment (jkosh44) [reply src/pgrepr/src/types.rs:94]: Oh, you're talking about the second part
> no more than 64 characters
Yeah, it's bytes not characters. I'll update this comment.

Comment (jkosh44) [reply src/pgrepr/src/types.rs]: Done

**Review by jkosh44 — APPROVED** [review]

**Pull request merged**